### PR TITLE
Security, Refactoring, and Improvements for TA grading

### LINKED ISSUE #1128: Refactor ElectronicGraderController::submitGrade() to use Gradeable::saveData()
Currently, when a TA submits a grade for a homework, the function submitGrade() puts all of the relevant data in an array then passes that to submitTAGrade() in DatabaseQueriesPostgresql.php. 

submitGrade() should be refactored to instead set the relevant data in the currently loaded gradeable model then call gradeable->saveData().

### PR DESCRIPTION
**Pull request by James-M-Anderson**
Closes #1128

Fixes several issues related to the Status, Details, and TA Grading Pages:
- Added checking of user's grading group against gradeable's minumum allowed grading group when accessing grading pages. Grading navigation page buttons are also removed when user is above the minimum grading group.
- Refactored ElectronicGraderController::submitGrade() to use Gradeable::saveData() when storing TA grades and comments in the database.
- Grade next student button and previous/next student buttons on the TA grading page now correspond to the order students are listed on the details page.
- Removed some dead code from HW reports and fixed bug which referred to "results_grade.txt", now called "grade.txt."
- Fixed indentation problems in ElectronicGraderView::statusPage() and fixed an html bug that caused the footer to be inside the content div.
- Moved Autograding results display code into new file AutogradingView. This view is called via renderTemplate() in both HomeworkView and ElectronicGraderView. 
- Visual changes to how hidden autograding testcases are displayed for TAs.

### CODE REVIEW

File: site/app/models/GradeableTestcase.php [hunk 1]
```diff
@@ -152,7 +152,7 @@ public function isExtraCredit() {
     }
     
     public function hasDetails() {
-      return (!$this->isHidden() && count($this->autochecks) > 0);
+      return (count($this->autochecks) > 0);
```
Comment (MasterOdin) [inline site/app/models/GradeableTestcase.php]: while you're modifying this model, can you add the `/** @property @var bool */` comment to both `$hidden` and `$extra_credit`, and then we can remove the `isHidden()` and `isExtraCredit()` functions (add them to the block of `@method ...` above the class name.

File: site/app/views/AutoGradingView.php [hunk 2]
```diff
@@ -0,0 +1,310 @@
...
+                            $visible = "hidden";
+                        }
+                    } else {
+                        $visible = "hidden";
+                    }
+                    $title .= $description;
+                    $return .= <<<HTML
+                <h4>{$title} <span onclick="openPopUp('{$title}', {$count}, {$autocheck_cnt}, 0)" style="visibility: {$visible}"> <i class="fa fa-window-restore" style="visibility: {$visible}; cursor: pointer;"></i></span></h4>
+                <div id="container_{$count}_{$autocheck_cnt}_0">
+                <link rel="stylesheet" type="text/css" href="{$this->core->getConfig()->getBaseUrl()}css/jquery-ui.min.css" />
```
Comment (MasterOdin) [inline site/app/views/AutoGradingView.php]: Why are these here? You're always on a page where most of these are already loaded on the page in the `GlobalView::header()` function (except maybe ta-grading.css, which I'd use `Output::addCss()` function to tell the system to load it on render in the <head> tag on page load.

Comment (MasterOdin) [inline site/app/views/AutoGradingView.php]: No wait, I remember why these are here, it's for the popup isn't it, since it loads the contents of the container `<div>` directly into the JS function?

File: site/app/controllers/grading/ElectronicGraderController.php [hunk 3]
```diff
@@ -132,65 +143,79 @@ public function submitGrade() {
...
+        }
+
         $now = new \DateTime('now', $this->core->getConfig()->getTimezone());
         $homeworkDate = $gradeable->getGradeStartDate();
         if ($now < $homeworkDate) {
             $_SESSION['messages']['error'][] = "Grading is not open yet for {$gradeable->getName()}";
             $this->core->redirect($this->core->buildUrl(array('component'=>'grading', 'page'=>'electronic', 'gradeable_id' => $gradeable_id)));
         }
+
+        if (isset($_POST['overwrite'])) $gradeable->setGrader($this->core->getUser());
```
Comment (bmcutler) [inline site/app/controllers/grading/ElectronicGraderController.php]: We still want the option to overwrite or not overwrite the original grader.  Sometimes the grade change is minor, and we still want the original grader to get the regrade request emails.

Comment (MasterOdin) [inline site/app/controllers/grading/ElectronicGraderController.php]: And we save that on line 164 for each component, which is what we're using to track who was the grader, not saving a grader for the whole gradeable.

Comment (James-M-Anderson) [inline site/app/controllers/grading/ElectronicGraderController.php]: HW reports and the overwrite grader checkbox on the TA grading page still use the grader stored in the gradeable model so this has to stay for now. I think Marisa is currently working on that.

File: site/app/views/AutoGradingView.php (hunk 2 above)
Comment (James-M-Anderson) [inline site/app/views/AutoGradingView.php]: The first four are for the popup, then I have to link ta-grading.css again because some things in there gets overwritten and mess up the rubric panels. Is there a better way to do this?

Comment (MasterOdin) [inline site/app/views/AutoGradingView.php]: Well, you almost never want to import stylesheets inside the body tag of a page, but on the other hand, we don't want to spend all the time remaking the diff stuff if we don't have to (especially for large file where this operation is expensive).

I think a better solution would be add a new parameter to openPopUp `$css` which you could pass in a list of all CSS that you want in the popup an then just dynamically add them in the function.
